include/eos/fitting/ceres_nonlinear.hpp

```diff
@@ -30,29 +33,46 @@
 #include "glm/gtc/quaternion.hpp"
 #include "glm/gtx/transform.hpp"
 
+#include "ceres/ceres.h"
 #include "ceres/cubic_interpolation.h"
+#include "ceres/problem.h"
 
+
+#ifndef EOS_CERES_USE_OPENCV
```

Comment:
We should actually use `eos::core::Image` now, since we got rid of all of OpenCV inside the eos library (except for the example app itself, where we use OpenCV to load/save images).
Then we can get rid of all these ifdef's :-)

Reply:
It is a good idea to have your own Image class, but there is a problem in the current implementation.

`ceres::Grid2D` [works](https://github.com/patrikhuber/eos/blob/8c6137ce4ae423e6b9dd231e61e8d9536a6815e7/include/eos/fitting/ceres_nonlinear.hpp#L358) with a raw pointer of data. Your class store values in `vector<std::uint8_t>`. It is possible to make `ptr()` member function that will use `reinterpret_cast` on it and return needed pointer, but it is not very obvious how to get the type of pointer to return. For example in case when `PixelType` is `double` we need to return just `double*`, but if `PixelType` is `Pixel<double, 3>` we need to take type from `Pixel` somehow.

There is no simple general solution, so I see just 2 ways how to solve it:
1) Just copy image while initialization and use it in `ImageCost` as a buffer.
2) Make a `template<typename T> T* ptr()` member function of `Image` which will reinterpret cast data to any type.

Can you suggest something better?